File: src/solar/solar-common/actor/util/ActorUtil.ts
import Actor from "@/Actor";
import DataBase from "@/DataBase";

/**
 * 演员通常最终作用于单位身上。（事实上所有游戏 所有数据 所有逻辑 最终的表现几乎都是在单位身上）
 * 优先使用此演员工具 不要把演员局限到技能或物品上 这样会导致演员不够统一 也不利于转换结构 比如吞噬这些 要保证无论是演员技能还是演员物品 都应该是一样的 所以优先考虑把逻辑代码写到演员基类里
 */
export default class ActorUtil {


    /**
     * 返回一个单位包含的所有演员  数据
     * 包括 演员单位 演员技能 演员物品 演员buff
     * 可以用来执行跟单位相关的通用逻辑  比如销毁单位时 或单位触发基础事件时 通用分发
     * @param unit
     */
    static getUnitAllActorList(unit: unit): Actor[] {
        if (!IsHandle(unit)) {
            return null;
        }
        let actorList: Actor[] = null;
        let solarData = DataBase.getUnitSolarData(unit, false);
        if (solarData != null) {
            //单位
            let actor = solarData._SL_solarActorUnit;
            if (actor != null) {
                actorList = [actor];
            }
            //技能
            let abilityTemplate = solarData._SL_abilityTemplate;
            if (abilityTemplate != null) {
                if (actorList == null) {
                    actorList = [];
                }
                for (let abilityTemplateKey in abilityTemplate) {
                    let actor = DataBase.getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)?._SL_solarActorAbility;
                    if (actor != null) {
                        actorList.push(actor)
                    }
                }
            }
            //演员buff
            let _SL_solarActorBuffSet: Actor[] = solarData._SL_solarActorBuffs;
            if (_SL_solarActorBuffSet) {
                for (let actor of _SL_solarActorBuffSet) {
                    if (actorList == null) {
                        actorList = [];
                    }
                    actorList.push(actor)
                }
            }
        }
        let invSize = UnitInventorySize(unit)
        for (let i = 0; i < invSize; i++) {
            let item = UnitItemInSlot(unit, i);
            if (IsHandle(item)) {
                let actor = DataBase.getItemSolarData(item, false)?._SL_solarActorItem;
                if (actor != null) {
                    if (actorList == null) {
                        actorList = [];
                    }
                    actorList.push(actor)
                }
            }
        }
        return actorList;
    }

    /**
     * 遍历单位身上的所有演员
     * @param unit
     * @param callback
     * @param clazz
     */
    static forUnitAllActorList(unit: unit, callback: (actor: Actor) => void, clazz?: string): void {
        let actorList = ActorUtil.getUnitAllActorList(unit);
        if (actorList == null) {
            return
        }
        for (let actor of actorList) {
            if (clazz == null || clazz == actor.get("class")) {
                callback(actor);
            }
        }
    }

    /**
     * 获取单位身上的所有演员 (指定演员类型 Class)
     * @param unit
     * @param clazz 同Class
     */
    static getUnitAllActorListByClass(unit: unit, clazz: string): Actor[] {
        let actorList = ActorUtil.getUnitAllActorList(unit);
        if (actorList == null) {
            return
        }
        let result: Actor[] = []
        for (let actor of actorList) {
            if (clazz == null || clazz == actor.get("class")) {
                result.push(actor);
            }
        }
        return result;
    }

    /**
     * 获取单位身上的所有演员 (指定演员类型 Class)
     * @param unit
     * @param clazz 同Class
     * @param kind
     * @param tag
     */
    static getUnitAllActorListAndWhere(unit: unit, clazz?: string, kind?: string, tag?: string): Actor[] {
        let actorList = ActorUtil.getUnitAllActorList(unit);
        if (actorList == null) {
            return
        }
        let result: Actor[] = []
        for (let actor of actorList) {
            if (clazz != null && clazz != actor.get("class")) {
                continue
            }
            if (kind != null && kind != actor.get("kind")) {
                continue
            }
            if (tag != null && tag != actor.get("tag")) {
                continue
            }
            result.push(actor);
        }
        return result;
    }

    /**
     * 获取单位身上所有buff 的属性值
     * @param unit
     */
    static getUnitAllActorAttributes(unit: unit): AppAttribute[] {
        let attributeArray: AppAttribute[] = null;
        let solarData = DataBase.getUnitSolarData(unit, false);
        if (solarData != null) {
            //单位
            let attribute: AppAttribute = solarData._SL_solarActorUnit?.attribute;
            if (attribute != null) {
                attributeArray = [attribute];
            }
            //技能
            let abilityTemplate = solarData._SL_abilityTemplate;
            if (abilityTemplate != null) {
                if (attributeArray == null) {
                    attributeArray = [];
                }
                for (let abilityTemplateKey in abilityTemplate) {
                    let attribute = DataBase.getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)?._SL_solarActorAbility?.attribute;
                    if (attribute != null) {
                        attributeArray.push(attribute)
                    }
                }
            }
            //演员buff
            let _SL_solarActorBuffSet: Actor[] = solarData._SL_solarActorBuffs;
            if (_SL_solarActorBuffSet) {
                for (let actorBuff of _SL_solarActorBuffSet) {
                    let attribute = actorBuff.attribute
                    if (attribute != null) {
                        if (attributeArray == null) {
                            attributeArray = [];
                        }
                        attributeArray.push(attribute)
                    }
                }
            }
        }
        let invSize = UnitInventorySize(unit)
        for (let i = 0; i < invSize; i++) {
            let item = UnitItemInSlot(unit, i);
            if (IsHandle(item)) {
                let attribute = DataBase.getItemSolarData(item, false)?._SL_solarActorItem?.attribute;
                if (attribute != null) {
                    if (attributeArray == null) {
                        attributeArray = [];
                    }
                    attributeArray.push(attribute)
                }
            }
        }
        return attributeArray;
    }

    /**
     * 如果单位是否持有某个类型的演员
     * @param unit
     * @param callBack
     * @param actorTypeId
     */
    static ifUnitHasActor(unit: unit, actorTypeId: string, callBack: (actor: Actor) => void): void {
        let unitAllActorList = ActorUtil.getUnitAllActorList(unit);
        if (unitAllActorList == null) {
            return;
        }
        for (let actor of unitAllActorList) {
            if (actor.actorTypeId == actorTypeId) {
                callBack(actor);
                return;
            }
        }
    }

    /**
     * 判断单位是否持有某个类型的演员(物品 技能 buff)
     * @param unit
     * @param actorTypeId
     */
    static isUnitHasActor(unit: unit, actorTypeId: string): boolean {
        let unitAllActorList = ActorUtil.getUnitAllActorList(unit);
        if (unitAllActorList == null) {
            return false;
        }
        for (let actor of unitAllActorList) {
            if (actor.actorTypeId == actorTypeId) {
                return true;
            }
        }
        return false
    }

    /**
     * 添加任意演员创建事件 监听回调
     */
    static addAnyActorCreatedListener(onActorCreatedListener: (actor: Actor) => void) {
        Actor._sl_anyActorCreatedListeners.push(onActorCreatedListener);
    }

    /**
     * 添加任意演员叠加层数改变事件 监听回调
     */
    static addAnyActorLevelChangeListener(onActorLevelChangeListener: (actor: Actor, stackAdd: number) => void) {
        Actor._sl_anyActorLevelChangeListeners.push(onActorLevelChangeListener);
    }

    /**
     * 添加任意演员销毁事件 监听回调
     */
    static addAnyActorDestroyListener(onActorDestroyListener: (actor: Actor) => void) {
        Actor._sl_anyActorDestroyListeners.push(onActorDestroyListener);
    }

}

File: scripts/solar/solar-common/actor/util/ActorUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 2,["9"] = 2,["10"] = 8,["11"] = 8,["12"] = 8,["14"] = 8,["15"] = 17,["16"] = 18,["17"] = 19,["19"] = 21,["20"] = 22,["21"] = 23,["22"] = 25,["23"] = 26,["24"] = 27,["26"] = 30,["27"] = 31,["28"] = 32,["29"] = 33,["31"] = 35,["32"] = 36,["33"] = 36,["34"] = 37,["35"] = 38,["39"] = 43,["40"] = 44,["41"] = 45,["42"] = 46,["43"] = 47,["45"] = 49,["49"] = 53,["51"] = 54,["52"] = 54,["53"] = 55,["54"] = 56,["55"] = 57,["56"] = 57,["57"] = 58,["58"] = 59,["59"] = 60,["61"] = 62,["64"] = 54,["67"] = 66,["68"] = 17,["69"] = 75,["70"] = 76,["71"] = 77,["74"] = 80,["75"] = 81,["76"] = 82,["79"] = 75,["80"] = 92,["81"] = 93,["82"] = 94,["85"] = 97,["86"] = 98,["87"] = 99,["88"] = 100,["91"] = 103,["92"] = 92,["93"] = 113,["94"] = 114,["95"] = 115,["98"] = 118,["99"] = 119,["101"] = 120,["102"] = 121,["104"] = 123,["105"] = 124,["107"] = 126,["108"] = 127,["110"] = 129,["114"] = 131,["115"] = 113,["116"] = 138,["117"] = 139,["118"] = 140,["119"] = 141,["120"] = 143,["122"] = 143,["124"] = 143,["125"] = 144,["126"] = 145,["128"] = 148,["129"] = 149,["130"] = 150,["131"] = 151,["133"] = 153,["134"] = 154,["135"] = 154,["137"] = 154,["139"] = 154,["140"] = 155,["141"] = 156,["145"] = 161,["146"] = 162,["147"] = 163,["148"] = 164,["149"] = 165,["150"] = 166,["151"] = 167,["153"] = 169,["158"] = 174,["160"] = 175,["161"] = 175,["162"] = 176,["163"] = 177,["164"] = 178,["165"] = 178,["167"] = 178,["169"] = 178,["170"] = 179,["171"] = 180,["172"] = 181,["174"] = 183,["177"] = 175,["180"] = 187,["181"] = 138,["182"] = 196,["183"] = 197,["184"] = 198,["187"] = 201,["188"] = 202,["189"] = 203,["193"] = 196,["194"] = 214,["195"] = 215,["196"] = 216,["197"] = 217,["199"] = 219,["200"] = 220,["201"] = 221,["204"] = 224,["205"] = 214,["206"] = 230,["207"] = 231,["208"] = 231,["209"] = 230,["210"] = 237,["211"] = 238,["212"] = 238,["213"] = 237,["214"] = 244,["215"] = 245,["216"] = 245,["217"] = 244});
local ____exports = {}
local ____Actor = require("solar.solar-common.actor.Actor")
local Actor = ____Actor.default
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
____exports.default = __TS__Class()
local ActorUtil = ____exports.default
ActorUtil.name = "ActorUtil"
function ActorUtil.prototype.____constructor(self)
end
function ActorUtil.getUnitAllActorList(self, unit)
    if not IsHandle(unit) then
        return nil
    end
    local actorList = nil
    local solarData = DataBase:getUnitSolarData(unit, false)
    if solarData ~= nil then
        local actor = solarData._SL_solarActorUnit
        if actor ~= nil then
            actorList = {actor}
        end
        local abilityTemplate = solarData._SL_abilityTemplate
        if abilityTemplate ~= nil then
            if actorList == nil then
                actorList = {}
            end
            for abilityTemplateKey in pairs(abilityTemplate) do
                local ____opt_0 = DataBase:getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)
                local actor = ____opt_0 and ____opt_0._SL_solarActorAbility
                if actor ~= nil then
                    actorList[#actorList + 1] = actor
                end
            end
        end
        local _SL_solarActorBuffSet = solarData._SL_solarActorBuffs
        if _SL_solarActorBuffSet then
            for ____, actor in ipairs(_SL_solarActorBuffSet) do
                if actorList == nil then
                    actorList = {}
                end
                actorList[#actorList + 1] = actor
            end
        end
    end
    local invSize = UnitInventorySize(unit)
    do
        local i = 0
        while i < invSize do
            local item = UnitItemInSlot(unit, i)
            if IsHandle(item) then
                local ____opt_2 = DataBase:getItemSolarData(item, false)
                local actor = ____opt_2 and ____opt_2._SL_solarActorItem
                if actor ~= nil then
                    if actorList == nil then
                        actorList = {}
                    end
                    actorList[#actorList + 1] = actor
                end
            end
            i = i + 1
        end
    end
    return actorList
end
function ActorUtil.forUnitAllActorList(self, unit, callback, clazz)
    local actorList = ____exports.default:getUnitAllActorList(unit)
    if actorList == nil then
        return
    end
    for ____, actor in ipairs(actorList) do
        if clazz == nil or clazz == actor:get("class") then
            callback(nil, actor)
        end
    end
end
function ActorUtil.getUnitAllActorListByClass(self, unit, clazz)
    local actorList = ____exports.default:getUnitAllActorList(unit)
    if actorList == nil then
        return
    end
    local result = {}
    for ____, actor in ipairs(actorList) do
        if clazz == nil or clazz == actor:get("class") then
            result[#result + 1] = actor
        end
    end
    return result
end
function ActorUtil.getUnitAllActorListAndWhere(self, unit, clazz, kind, tag)
    local actorList = ____exports.default:getUnitAllActorList(unit)
    if actorList == nil then
        return
    end
    local result = {}
    for ____, actor in ipairs(actorList) do
        do
            if clazz ~= nil and clazz ~= actor:get("class") then
                goto __continue32
            end
            if kind ~= nil and kind ~= actor:get("kind") then
                goto __continue32
            end
            if tag ~= nil and tag ~= actor:get("tag") then
                goto __continue32
            end
            result[#result + 1] = actor
        end
        ::__continue32::
    end
    return result
end
function ActorUtil.getUnitAllActorAttributes(self, unit)
    local attributeArray = nil
    local solarData = DataBase:getUnitSolarData(unit, false)
    if solarData ~= nil then
        local ____opt_4 = solarData._SL_solarActorUnit
        if ____opt_4 ~= nil then
            ____opt_4 = ____opt_4.attribute
        end
        local attribute = ____opt_4
        if attribute ~= nil then
            attributeArray = {attribute}
        end
        local abilityTemplate = solarData._SL_abilityTemplate
        if abilityTemplate ~= nil then
            if attributeArray == nil then
                attributeArray = {}
            end
            for abilityTemplateKey in pairs(abilityTemplate) do
                local ____opt_8 = DataBase:getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)
                local ____opt_6 = ____opt_8 and ____opt_8._SL_solarActorAbility
                if ____opt_6 ~= nil then
                    ____opt_6 = ____opt_6.attribute
                end
                local attribute = ____opt_6
                if attribute ~= nil then
                    attributeArray[#attributeArray + 1] = attribute
                end
            end
        end
        local _SL_solarActorBuffSet = solarData._SL_solarActorBuffs
        if _SL_solarActorBuffSet then
            for ____, actorBuff in ipairs(_SL_solarActorBuffSet) do
                local attribute = actorBuff.attribute
                if attribute ~= nil then
                    if attributeArray == nil then
                        attributeArray = {}
                    end
                    attributeArray[#attributeArray + 1] = attribute
                end
            end
        end
    end
    local invSize = UnitInventorySize(unit)
    do
        local i = 0
        while i < invSize do
            local item = UnitItemInSlot(unit, i)
            if IsHandle(item) then
                local ____opt_12 = DataBase:getItemSolarData(item, false)
                local ____opt_10 = ____opt_12 and ____opt_12._SL_solarActorItem
                if ____opt_10 ~= nil then
                    ____opt_10 = ____opt_10.attribute
                end
                local attribute = ____opt_10
                if attribute ~= nil then
                    if attributeArray == nil then
                        attributeArray = {}
                    end
                    attributeArray[#attributeArray + 1] = attribute
                end
            end
            i = i + 1
        end
    end
    return attributeArray
end
function ActorUtil.ifUnitHasActor(self, unit, actorTypeId, callBack)
    local unitAllActorList = ____exports.default:getUnitAllActorList(unit)
    if unitAllActorList == nil then
        return
    end
    for ____, actor in ipairs(unitAllActorList) do
        if actor.actorTypeId == actorTypeId then
            callBack(nil, actor)
            return
        end
    end
end
function ActorUtil.isUnitHasActor(self, unit, actorTypeId)
    local unitAllActorList = ____exports.default:getUnitAllActorList(unit)
    if unitAllActorList == nil then
        return false
    end
    for ____, actor in ipairs(unitAllActorList) do
        if actor.actorTypeId == actorTypeId then
            return true
        end
    end
    return false
end
function ActorUtil.addAnyActorCreatedListener(self, onActorCreatedListener)
    local ____Actor__sl_anyActorCreatedListeners_14 = Actor._sl_anyActorCreatedListeners
    ____Actor__sl_anyActorCreatedListeners_14[#____Actor__sl_anyActorCreatedListeners_14 + 1] = onActorCreatedListener
end
function ActorUtil.addAnyActorLevelChangeListener(self, onActorLevelChangeListener)
    local ____Actor__sl_anyActorLevelChangeListeners_15 = Actor._sl_anyActorLevelChangeListeners
    ____Actor__sl_anyActorLevelChangeListeners_15[#____Actor__sl_anyActorLevelChangeListeners_15 + 1] = onActorLevelChangeListener
end
function ActorUtil.addAnyActorDestroyListener(self, onActorDestroyListener)
    local ____Actor__sl_anyActorDestroyListeners_16 = Actor._sl_anyActorDestroyListeners
    ____Actor__sl_anyActorDestroyListeners_16[#____Actor__sl_anyActorDestroyListeners_16 + 1] = onActorDestroyListener
end
return ____exports
